### PR TITLE
Keys are compared in BucketSortPipelineAggregation so making key type…

### PR DESCRIPTION
**Pull request by ItamarBenjamin**
… (ArrayMap) implement Comparable. Maps are compared using the entry set's iterator so ordered maps order is maintain. For each entry first comparing key then value. Assuming all keys are strings. When comparing entries' values if type is not identical and\or type not implementing Comparable, throwing exception. Not implementing equals() and hashCode() functions as parent's ones are sufficient. Tests included.

### CODE REVIEW

**Comment by elasticcla** [comment]
Hi @ItamarBenjamin, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in yout Git [commit](https://github.com/elastic/elasticsearch/pull/36407.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/InternalComposite.java [hunk 1]
```diff
@@ -392,16 +393,17 @@ static Object formatObject(Object obj, DocValueFormat format) {
         return obj;
     }
 
-    private static class ArrayMap extends AbstractMap<String, Object> {
+    static class ArrayMap extends AbstractMap<String, Object> implements Comparable<Map<String,Object>> {
         final List<String> keys;
-        final List<DocValueFormat> formats;
         final Object[] values;
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/InternalComposite.java]: The `values` are already comparable so you could change the type to `Comparable<?>` here and then use `values[i].compareTo` ?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/InternalComposite.java [hunk 2]
```diff
@@ -392,16 +393,17 @@ static Object formatObject(Object obj, DocValueFormat format) {
...
 
         ArrayMap(List<String> keys, List<DocValueFormat> formats, Object[] values) {
             assert keys.size() == values.length && keys.size() == formats.size();
             this.keys = keys;
-            this.formats = formats;
-            this.values = values;
+            this.values = new Object[values.length];
+            for (int i = 0; i < values.length; i++) {
+                this.values[i] = formatObject(values[i], formats.get(i));
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/InternalComposite.java]: If you want a comparaison of the raw key you cannot perform the formatting in the ctr. Depending on the requested format the value can be formatted to a string which will not preserve the initial order (dates for instance).

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/InternalComposite.java [hunk 3]
```diff
@@ -447,5 +449,63 @@ public int size() {
...
+        public int compareTo(Map<String,Object> that) {
+            if (that == this) {
+                return 0;
+            }
+            Iterator<Entry<String, Object>> thisIterator = this.entrySet().iterator();
+            Iterator<Entry<String, Object>> thatIterator = that.entrySet().iterator();
+            while (thisIterator.hasNext() && thatIterator.hasNext()) {
+                Entry<String, Object> thisNext = thisIterator.next();
+                Entry<String, Object> thatNext = thatIterator.next();
+                int compare = compareInternal(thisNext.getKey(),thatNext.getKey(),String::compareTo);
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/InternalComposite.java]: `Objects.compare` ?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/InternalComposite.java [hunk 4]
```diff
@@ -392,16 +393,17 @@ static Object formatObject(Object obj, DocValueFormat format) {
         return obj;
     }
 
-    private static class ArrayMap extends AbstractMap<String, Object> {
+    static class ArrayMap extends AbstractMap<String, Object> implements Comparable<Map<String,Object>> {
         final List<String> keys;
-        final List<DocValueFormat> formats;
         final Object[] values;
 
         ArrayMap(List<String> keys, List<DocValueFormat> formats, Object[] values) {
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/InternalComposite.java]: `ArrayMap(List<String> keys, List<DocValueFormat> formats, Comparable<?>[] values) {`

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/InternalComposite.java [hunk 5]
```diff
@@ -447,5 +449,63 @@ public int size() {
...
+            }
+            throw new IllegalStateException("expecting values types to implement java.lang.Comparable but got: " + a.getClass());
+        }
+
+        @Override
+        public int compareTo(Map<String,Object> that) {
+            if (that == this) {
+                return 0;
+            }
+            Iterator<Entry<String, Object>> thisIterator = this.entrySet().iterator();
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/InternalComposite.java]: The iterator returns the formatted values so you should simply iterate directly on the `values` array ?

**Review by jimczi — CHANGES_REQUESTED** [review]
Thanks @ItamarBenjamin , the change looks good. I left some comments regarding the type of values that you need to compare.

**Comment by ItamarBenjamin** [comment]
Hi @jimczi, Thanks for your comments. please review the new commit.

**Comment by jimczi** [comment]
@elasticmachine test this please

**Comment by jimczi** [comment]
@elasticmachine test this please

**Comment by jimczi** [comment]
@elasticmachine test this please

**Comment by jimczi** [comment]
@elasticmachine test this please

**Comment by jimczi** [comment]
run gradle build tests 2

**Review by jimczi — APPROVED** [review]
Sorry for the delay @ItamarBenjamin , the change looks good and tests pass so I'll merge

**Pull request merged**